### PR TITLE
All things log-in!

### PR DESCRIPTION
**Pull request by ryan-garay89**
This PR adds full functionality for logging in and out. I wrote this code fairly quickly so please make sure it looks okay. The login flow implemented here **does not** verify that a user is an admin or counselor, but rather assumes that the user is a student. We will have to implement functionality for verifying user types at a later time.

I ask you to please check out this branch locally and do some testing / play around with it and see if you find any bugs. Once it's ready to be merged we will update the redirect_uri on Google.

### CODE REVIEW

File: app/views/pages/index.html.erb [hunk 1]
```diff
@@ -9,24 +9,22 @@
...
       <h3 class="display-5 name-display mt-3"><%= @name.html_safe %></h3>
       <% if @user_type.include? 'Admin' %>
         <%= link_to "Go to admin dashboard".html_safe, admins_path, class: "btn btn-primary mt-1 btn-lg px-4 gap-3" %>
-      <% end %>
-      <% if @user_type.include? 'Staff' %>
+      <% elsif @user_type.include? 'Staff' %>
         <%# pending %>
-      <% end %>
-      <% if @user_type.include? 'Student' %>
+      <% else %>
```
Comment (ryan-garay89) [inline app/views/pages/index.html.erb:24]: for now, at least before admins & counselors are fully functional, let's assume a user is a student if they have no other type

File: app/controllers/user_controller.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    end
+    first_name = params[:user][:first_name]
+    last_name = params[:user][:last_name]
+    email = params[:user][:email]
+    sid = params[:user][:sid]
+    if first_name.nil? or first_name.empty? or last_name.nil? or last_name.empty? or email.nil? or email.empty? or (sid.nil? or sid.empty? and @user.sid.nil?)
+      redirect_to login_confirm_path, flash: { :error => "Something went wrong, please try again" }
+      return
+    end
+    if sid.length != 10
```
Comment (KLongLo) [inline app/controllers/user_controller.rb]: Here should check if sid is not nil, otherwise sid.length will raise an error

**Review by KLongLo — CHANGES_REQUESTED** [review]

File: app/controllers/user_controller.rb (hunk 2 above)
Comment (KLongLo) [reply app/controllers/user_controller.rb]: Other things look good to me :)

Comment (ryan-garay89) [reply app/controllers/user_controller.rb]: will do! thank you

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -4,14 +4,22 @@ def googleAuth
         # Get access tokens from the google server
         access_token = request.env["omniauth.auth"]
         user = User.from_omniauth(access_token)
-        log_in(user)
         # Access_token is used to authenticate request made from the rails application to the google server
         user.google_token = access_token.credentials.token
         # Refresh_token to request new access_token
         # Note: Refresh_token is only sent once during the first request
         refresh_token = access_token.credentials.refresh_token
         user.google_refresh_token = refresh_token if refresh_token.present?
```
Comment (KLongLo) [inline app/controllers/sessions_controller.rb:13]: Do we really need these lines?

**Review by KLongLo — CHANGES_REQUESTED** [review]

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (ryan-garay89) [reply app/controllers/sessions_controller.rb:13]: I think they allow google to request new logins for security purposes (the google auth expires) -- maybe @michellekroll knows more?

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -3,15 +3,28 @@ class SessionsController < ApplicationController
...
-        user.save
-        redirect_to root_path
+        if user.save
+            if existing_user
+                redirect_to root_path, flash: { :success => "Success! You've been logged-in!" }
+            else
+                session[:current_user_id] = user.id
+                redirect_to login_confirm_path
```
Comment (ryan-garay89) [inline app/controllers/sessions_controller.rb:20]: only new users should be redirected to verify their information

File: app/controllers/user_controller.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    @user = User.find_by(id: session[:current_user_id])
+    if @user.nil?
+      redirect_to login_path, flash: { :error => "Please log-in!" }
+      return
+    end
+    first_name = params[:user][:first_name]
+    last_name = params[:user][:last_name]
+    email = params[:user][:email]
+    sid = params[:user][:sid]
+    if sid.blank? or sid.length != 10
```
Comment (ryan-garay89) [inline app/controllers/user_controller.rb]: I updated the checks for first_name, last_name, etc. to be validated in the User model. Now I only need to check that the sid entered isn't blank (nil or empty) and that the length is valid.

File: app/views/login/confirm.html.erb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  </div>
+  <div class="px-4 py-3 my-5 text-center">
+    <h2 class="display-5 fw-bold pb-1">Welcome!</h2>
+    <h3 class="display-5 pb-2">Please Verify All the information below.</h3>
+      <%= form_with(model: @user, local: true) do |f| %>
+        <div class="row mb-3">
+          <div class="col-md-6 offset-md-3">
+            <div class="col-md-6 offset-md-3">
+              <%= f.label :first_name, class: "form-label" %>
+              <%= f.text_field :first_name, value: @user.first_name, :required => true, class: "form-control" %>
```
Comment (ryan-garay89) [inline app/views/login/confirm.html.erb:15]: added :required fields to text inputs on the form to validate on the frontend that fields aren't empty

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (michellekroll) [reply app/controllers/sessions_controller.rb:13]: I am not sure but I can look into it. I believe the functionality we have right now will not do that though.

File: app/controllers/sessions_controller.rb (hunk 4 above)
Comment (KLongLo) [reply app/controllers/sessions_controller.rb:20]: Should we do user.save only if existing_user does not exist?

**Comment by KLongLo** [comment]
When the same user logs in the second time, the landing page will look like this, which doesn't look right 🤔:
image


edit: I think the problem occurs because the program doesn't set `session[:current_session_id]` if the logging in user is an existing user?

**Comment by KLongLo** [comment]
Also, I think it is better to write some integration tests for the login features before merging to the main branch?

**Comment by ryan-garay89** [comment]
> When the same user logs in the second time, the landing page will look like this, which doesn't look right 🤔:
> 
> image
> 
> 
> 
> 
> 
> edit: I think the problem occurs because the program doesn't set `session[:current_session_id]` if the logging in user is an existing user?

yes, my bad! I will fix

**Comment by KLongLo** [comment]
The confirm.html.erb is actually updating the attributes of an user. Should make a new method in user controller (and its corresponding view) for handling the attribute updates?

**Comment by ryan-garay89** [comment]
> The confirm.html.erb is actually updating the attributes of an user. Should make a new method in user controller (and its corresponding view) for handling the attribute updates?

the confirm.html.erb doesn't actually change or interact with the record as far as I know. And the attribute updates are handled in the user controller (in the update method)

**Review by KLongLo — APPROVED** [review]
LGTM!!

**Pull request merged**